File: app/src/main/java/com/kymjs/coroutine/MainActivity.java
package com.kymjs.coroutine;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import com.kymjs.coroutine.databinding.ActivityMainBinding;
import com.kymjs.thread.LiteThread;

public class MainActivity extends AppCompatActivity {

    static {
        System.loadLibrary("coroutine");
    }

    private ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        TextView tv = binding.sampleText;
        tv.setText("test ndk");

        binding.button1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new TestThread().start();
            }
        });
        binding.button2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new TestThread().start();
            }
        });
    }

    static class TestThread extends LiteThread {
        @Override
        public void run() {
            super.run();
            System.out.println(Thread.currentThread().getName());
            System.out.println("test in java run method");
            int i = 0;
            while (true) {
                System.out.println("i=" + i++);
                sleep(1000, threadId);
                System.out.println("当前中断状态:" + isInterrupt());
                if (i > 5) {
                    if (isInterrupt()) {
                        resume();
                        System.out.println("恢复");
                    } else {
                        interrupt();
                        System.out.println("中断");
                    }
                    i = 0;
                }
            }
        }
    }
}

File: app/src/main/java/com/kymjs/coroutine/MainActivity.kt
package com.kymjs.coroutine

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.TextView
import com.kymjs.coroutine.MainActivity.TestThread
import com.kymjs.coroutine.MainActivity.TestCoroutine
import com.kymjs.thread.LiteThread
import com.kymjs.coroutine.Coroutine
import com.kymjs.coroutine.databinding.ActivityMainBinding
import java.lang.Exception
import java.net.URL

class MainActivity : AppCompatActivity() {
    companion object {
        init {
            System.loadLibrary("coroutine")
        }
    }

    private var binding: ActivityMainBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        val tv = binding!!.sampleText
        tv.text = "test ndk"
        binding!!.button1.setOnClickListener { TestThread().start() }
        binding!!.button2.setOnClickListener { TestCoroutine().async() }
        binding!!.button3.setOnClickListener { TestCoroutine(2).async() }
    }

    internal class TestThread : LiteThread() {
        override fun run() {
            super.run()
            println(Thread.currentThread().name)
            println("test in java despatch method")
            var i = 0
            while (true) {
                println("i=" + i++)
                sleep(1000, threadId)
                println("当前中断状态:$isInterrupt")
                if (i > 5) {
                    if (isInterrupt) {
                        resume()
                        println("恢复")
                    } else {
                        interrupt()
                        println("中断")
                    }
                    i = 0
                }
            }
        }
    }

    internal class TestCoroutine : Coroutine {
        constructor() : super() {}
        constructor(id: Long) : super(id) {}

        override fun run(): Any {
            super.run()
            try {
                return URL("https://baidu.com").readText()
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return "test in java"
        }

        override fun <T> onAwait(result: T) {
            super.onAwait(result)
            println("========output::$result")
        }
    }
}